### ISSUE TITLE
getAppFilename returns (deleted) in the path

### ISSUE THREAD

**Issue by dom96** [nim-lang/Nim#15079]
### Example
```nim
import os
while true:
  echo(getAppFilename())
  sleep(1000)
```

Run this and then recompile it. (This is on Linux/WSL)

### Current Output

```
/file/path/to/exe
/file/path/to/exe
/file/path/to/exe (deleted)
```

### Expected Output
```
/file/path/to/exe
/file/path/to/exe
/file/path/to/exe
```

### Additional Information
```
$ nim -v
Nim Compiler Version 1.0.6
```

**Comment by ghost** [comment]
Proposed solution: handle it the same way it's handled for macOS/Windows and document that (return an emptry string on error)

**Comment by Araq** [comment]
Not our bug, we read ``/proc/self/exe``, nothing more, nothing less.

**Comment by dom96** [comment]
I respectfully disagree.

**Comment by alaviss** [comment]
> Proposed solution: handle it the same way it's handled for macOS/Windows and document that (return an emptry string on error)

It's not an error, the ` (deleted)` suffix is added if the link target no longer exist (this quirk applies to "special" file system like `procfs` only iirc). There isn't much that we can do about this since there aren't any way to atomically get both the file name and its existence status.

**Comment by alaviss** [comment]
We can try to "avoid" this by querying this data at the program start, though if the executable is deleted before the initialization code is run then we're out of luck :P

Of course, this also come with issues like the file name won't be updated if the file is renamed.

**Comment by alaviss** [comment]
Note that if this happens without the file being deleted, then it's WSL bug and not ours.

**Comment by Araq** [comment]
Respectfully disagreeing is not a reason for re-opening, you need to justify it somehow. And we don't fix OS bugs, we cannot.

**Comment by dom96** [comment]
I would have thought that it would be obvious, but apparently not.

We need to at the very least document this behaviour. ``getAppFilename`` is not called ``readProcSelfExe`` so why are you closing this with the excuse that it's an OS bug? It's as if you created a ``getCPU`` proc that executed ``uname -p`` and then when someone complained that it doesn't always return a correct processor name you blamed the OS.

**Comment by Araq** [comment]
Well yes, it's really not clear what and how to "fix" the bug. Are we supposed to parse "/proc/self/exe" ? But what if your exe is named "foo (deleted)" ? Your OS is buggy.

**Comment by Araq** [comment]
> It's as if you created a getCPU proc that executed uname -p and then when someone complained that it doesn't always return a correct processor name you blamed the OS.

Exactly, now what should we do? Ask the BIOS instead of the OS? Never mind that user level programs cannot do that...

**Comment by dom96** [comment]
I don't know what you should do. But closing this issue and pretending it's not your problem is not a solution I am okay with.

**Comment by Araq** [comment]
It's indeed not my problem. Adding " (deleted)" to strings returned by APIs nilly-willy is insanity. But ok, we can leave the issue open as "won't fix" for a year instead, I'm sure it'll help.

**Comment by timotheecour** [comment]
how about combining this: https://stackoverflow.com/a/39001204/1426932
with this: https://stackoverflow.com/a/1189582/1426932

```nim
int fd = open("/proc/self/exe", O_RDONLY);
use readlink on /proc/self/fd/NNN
```
> This will give you the name of the file as it was when it was opened — however, if the file was moved or deleted since then, it may no longer be accurate (although Linux can track renames in some cases). To verify, stat the filename given and fstat the fd you have, and make sure st_dev and st_ino are the same.
[...]

other than that, a flag `splitDeleted` could be added for linux

**Comment by alaviss** [comment]
> how about combining this: https://stackoverflow.com/a/39001204/1426932
> with this: https://stackoverflow.com/a/1189582/1426932
> 
> ```nim
> int fd = open("/proc/self/exe", O_RDONLY);
> use readlink on /proc/self/fd/NNN
> ```
> 
> > This will give you the name of the file as it was when it was opened — however, if the file was moved or deleted since then, it may no longer be accurate (although Linux can track renames in some cases). To verify, stat the filename given and fstat the fd you have, and make sure st_dev and st_ino are the same.
> > [...]

That's just `readlink` with extra steps, it won't make any difference.

> other than that, a flag `splitDeleted` could be added for linux

I wish we could just detect when to split `(deleted)`, but the truth of the matter is that no matter how you do it you end up with a TOCTOU problem. It's better to just not do anything and document the issue instead.
